Math/NumberTheory/Moduli/Sqrt.hs

```diff
@@ -10,33 +10,83 @@
 --
 
 {-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE ViewPatterns #-}
 {-# LANGUAGE CPP          #-}
 
 module Math.NumberTheory.Moduli.Sqrt
-  ( sqrtModP
-  , sqrtModPList
-  , sqrtModP'
-  , tonelliShanks
-  , sqrtModPP
-  , sqrtModPPList
+  ( -- Constructor and @'unPrime'@ are exported in order to do pattern-matching.
+    FieldCharacteristic(..)
+  , unPrime
+  , toFieldCharacteristic
+  , sqrtModList
+  , sqrtModMaybe
+  , sqrtModExact
   , sqrtModF
   , sqrtModFList
   ) where
 
 import Control.Monad (liftM2)
 import Data.Bits
+import Data.Coerce
 import Data.List (nub)
 import GHC.Integer.GMP.Internals
 
 import Math.NumberTheory.Moduli.Chinese
 import Math.NumberTheory.Moduli.Jacobi
+import Math.NumberTheory.Powers.General (highestPower)
+import Math.NumberTheory.Primes.Types
 import Math.NumberTheory.Primes.Sieve (sieveFrom)
+import Math.NumberTheory.Primes.Testing.Probabilistic (isPrime)
 import Math.NumberTheory.Utils (shiftToOddCount, splitOff)
 
+import Math.NumberTheory.Utils.FromIntegral
+
+-- | Extended flag structure to check if square root is computed in finite field.
+data FieldCharacteristic = FieldCharacteristic (Prime Integer) Int
+  deriving Show
+
+-- | Duplicate functions from @'instance UniqueFactorisation Integer'@. They cannot be imported because of cycle imports.
+unPrime :: PrimeNat -> Integer
+unPrime = coerce naturalToInteger
+
+primeF :: Integer -> PrimeNat
+primeF = coerce integerToNatural
+
+-- | Check if value is power of some prime.
+checkPrimePower :: Integer -> Maybe (Prime Integer, Int)
+checkPrimePower p
+  -- FIXME: @'mapFst'@ function presents in @'Data.Monoid.State'@
```

Comment:
Is `mapFst` the same to `Control.Arrow.first`?

Reply:
Ah. Yes